servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -164,29 +189,30 @@ public void onNext(final ServiceDiscovererEvent<ResolvedAddress> event) {
                         final List<Host<ResolvedAddress, C>> oldHostsTyped = (List<Host<ResolvedAddress, C>>) oldHosts;
                         if (event.isAvailable()) {
                             if (oldHostsTyped.isEmpty()) {
-                                return singletonList(new Host<>(addr));
+                                return singletonList(createHost(addr));
                             }
-                            final List<Host<ResolvedAddress, C>> newHosts = new ArrayList<>(oldHostsTyped.size() + 1);
+                            final List<Host<ResolvedAddress, C>> newHosts =
+                                    new ArrayList<>(oldHostsTyped.size() + 1);
                             newHosts.addAll(oldHostsTyped);
-                            newHosts.add(new Host<>(addr));
+                            newHosts.add(createHost(addr));
                             return newHosts;
                         } else if (oldHostsTyped.isEmpty()) {
                             return emptyList();
                         } else {
-                            final List<Host<ResolvedAddress, C>> newHosts = new ArrayList<>(oldHostsTyped.size() - 1);
-                            for (int i = 0; i < oldHostsTyped.size(); ++i) {
-                                final Host<ResolvedAddress, C> host = oldHostsTyped.get(i);
-                                if (host.address.equals(addr)) {
-                                    host.markInactive();
-                                    for (int x = i + 1; x < oldHostsTyped.size(); ++x) {
-                                        newHosts.add(oldHostsTyped.get(x));
+                            if (eagerConnectionShutdown) {
+                                return removeFromHostsList(oldHosts, addr);
+                            } else {
+                                for (Host<ResolvedAddress, C> host : oldHostsTyped) {
+                                    if (host.address.equals(addr)) {
+                                        host.isExpired = true;
+                                        if (host.connections.length == 0) {
+                                            host.closeAsyncGracefully().subscribe();
```

Comment:
I think this is still racy. Because there is a gap between checking the length and swapping the `connections` with `CLOSED_ARRAY` inside `doClose`, another thread that creates a new connection can add it to the `connections` array and return `true` from `addConnection`. It will mean that the connection is selected for request execution. Due to offloading, `closeAsyncGracefully()` may close the connection before the meta-data are taken by event-loop, and the state is updated inside `RequestResponseCloseHandler`. As the result, request will fail because connection is already closed.

Consider holding a ref to `host.connections` in a local variable. Then check the length and if it's 0, try to CAS with `CLOSED_ARRAY`. Do `closeAsyncGracefully()` only if CAS succeeded. Actually, we can do `closeAsync()` in this case.

Another approach can be to CAS `EMPTY_ARRAY` with `CLOSED_ARRAY`, without length check. But it will require changing `connection.onClose().beforeFinally` to always set connections to  `EMPTY_ARRAY` if it handles closure of the last connection when the host is not expired.